README.md
# hdx-data-extraction-ahcd
------
Code to convert CDC's Ambulatory Health Care Data (AHCD) (NAMCS and NHAMCS) in human readable form.
https://www.cdc.gov/nchs/ahcd/about_ahcd.htm.
  - # NAMCS
    The National Ambulatory Medical Care Survey (NAMCS) is a national survey designed to meet the need for objective, reliable information about the provision and use of ambulatory medical care services in the United States. Findings are based on a sample of visits to nonfederally employed office-based physicians who are primarily engaged in direct patient care.
  - # NHAMCS
    The National Hospital Ambulatory Medical Care Survey (NHAMCS) is designed to collect data on the utilization and provision of ambulatory care services in hospital emergency and outpatient departments, and in ambulatory surgery centers.

# Code Structure
>   **hdx_ahcd** serves as base directory.
```sh
hdx_ahcd
├── api.py
├── controllers
│   ├── __init__.py
│   ├── namcs_converter.py
│   ├── namcs_extractor.py
│   └── namcs_processors.py
├── helpers
│   ├── functions.py
│   └── __init__.py
├── __init__.py
├── mappers
│   ├── functions.py
│   ├── __init__.py
│   └── years.py
├── namcs
│   ├── config.py
│   ├── constants.py
│   ├── enums.py
│   └── __init__.py
├── scripts
│   ├── __init__.py
│   └── namcs_validators.py
└── utils
    ├── context.py
    ├── decorators.py
    ├── exceptions.py
    ├── __init__.py
    └── utils.py
```
* api - API to process NAMCS dataset file(s).
* controllers
    - namcs_extractor.py - Download and extract public NAMCS data.
    - namcs_converter.py - Process and convert NAMCS data in human readable form.
    - namcs_processors - Provide common entry point for execution.
* helpers - Various methods for manipulating dataset and it's details.
* mappers
    - helpers - Methods to translate raw data from dataset to human readable format.
    - years - Year wise NAMCS details like fields, field location, length etc.
* namcs - Contains configurable parameters and constants.
* scripts
    - namcs_validators - Validation of dataset and parameters provided while invoking script namcs_processors.
* utils - Contains useful decorators, context managers etc.
* namcs_test - Script to perform regression for all namcs year(DEV purpose only).
### Supported fields
-----
- date_of_visit - Patient date of visit.
- date_of_birth - Patient date of birth.
- year_of_visit - Patient year of visit.
- year_of_birth - Patient year of birth.
- month_of_visit - Patient month of visit.
- month_of_birth - Patient month of birth.
- patient_age - Patient age in days.
- gender - Patient gender.
- physician_diagnoses - ICD-9-CM code (International Classification of Diseases, 9th Revision, Clinical Modification) for Diagnostic information
- visit_weight - The "patient visit weight" is a vital component in the
process of producing national estimates from sample data,
and its use should be   clearly understood by all micro-data file
users. The statistics contained on the micro-data
file reflect data concerning only a sample of
patient visits, not a complete count of all the
visits that occurred in the United States. Each
record on the data file represents one
visit in the sample of 27,369 visits. In order to obtain
national estimates from the sample,
each record is assigned an inflation factor called the
"patient visit weight."

### Installation
-----
Currently supported python version 3.6.x,
To check python version
```sh
python --version
```
Ensure pip, setuptools, and wheel are up to date
```sh
python -m pip install --upgrade pip setuptools wheel
```
If you have local copy of this repo and want to install directly from it.
```sh
pip install ${PATH_FOR_hdx-data-extraction-ahcd_REPO}
```
Similarly you can execute setup file
```sh
python3 ${PATH_FOR_hdx-data-extraction-ahcd_REPO}/setup.py install
```
for example:
```sh
pip install /var/tmp/hdx-data-extraction-ahcd/
```
and

```sh
python3 /var/tmp/hdx-data-extraction-ahcd/setup.py install
```
You can also use pip directly for Installation.
```sh
pip install hdx_ahcd
```
-----
### Usage
-----
```sh
>>> import hdx_ahcd
>>> from hdx_ahcd.api import get_cleaned_data_by_year
>>> import pprint
>>> pp =pprint.PrettyPrinter(indent=4)
```
> Case 1: Downloading NAMCS data for single year (say, 1973). If the file is
        already present in the downloaded_files then process the downloaded
        file.
```sh
>>> gen =  get_cleaned_data_by_year(year=1973)
INFO:hdx_ahcd:Downloading file:ftp://ftp.cdc.gov
/pub/Health_Statistics/NCHS/namcs_public_use_files/namcs73.exe for year:1973
>>> pp.pprint(gen)
defaultdict(<class 'dict'>,
        {   1973: {   'generator': <generator object get_generator_by_year at 0x7fe4b6480150>,
                      'source_file_info': {
                      'url': 'ftp://ftp.cdc.gov/pub/Health_Statistics/NCHS/namcs_public_use_files/namcs73.exe',
                      'year': '73',
                      'zip_file_name': 'namcs73.exe'}}})
```
> Case 2: Downloading NAMCS data for multiple years (say, 1973 and 1975).
        If the file is already present in the downloaded_files then process the
        downloaded file.
```sh
>>> gen =  get_cleaned_data_by_year(year= (1973,1975))
INFO:hdx_ahcd:Downloading file:ftp://ftp.cdc.gov
/pub/Health_Statistics/NCHS/namcs_public_use_files/namcs75.exe for year:1975
>>> pp.pprint(gen)
defaultdict(<class 'dict'>,
            {   1973: {   'generator': <generator object get_generator_by_year at 0x7fe4b5fa9e08>,
                          'source_file_info': {
                          'url': 'ftp://ftp.cdc.gov/pub/Health_Statistics/NCHS/namcs_public_use_files/namcs73.exe',
                          'year': '73',
                          'zip_file_name': 'namcs73.exe'}},
                1975: {   'generator': <generator object
                get_generator_by_year at 0x7fe4b45e7e60>,
                          'source_file_info': {
                          'url': 'ftp://ftp.cdc.gov/pub/Health_Statistics/NCHS/namcs_public_use_files/namcs75.exe',
                          'year': '75',
                          'zip_file_name': 'namcs75.exe'}}})
```
> Iterating over generator
```sh
>>> pp.pprint(next(gen.get(1973).get('generator')))
{   'age': 22889,
    'month_of_visit': 'June',
    'patient_visit_weight': 13479.0,
    'physician_diagnoses': ['470.0', 'V03.2', ''],
    'sex': 'Female',
    'source_file_ID': '1973_NAMCS',
    'source_file_row': 1,
    'year_of_visit': '1973'}
>>> pp.pprint(next(gen.get(1975).get('generator')))
{   'age': 14610,
    'month_of_visit': 'April',
    'patient_visit_weight': 3722.0,
    'physician_diagnoses': ['492.0', '', ''],
    'sex': 'Male',
    'source_file_ID': '1975_NAMCS',
    'source_file_row': 1,
    'year_of_visit': '1975'}
```
> Case 3: Forcefully download and then process the NAMCS data for multiple
        years (say, 1973 and 1975).
```sh
>>> gen =  get_cleaned_data_by_year(year= (1973,1975), force_download=True)
INFO:hdx_ahcd:Downloading file:ftp://ftp.cdc.gov/pub/Health_Statistics/NCHS/namcs_public_use_files/namcs73.exe for year:1973
INFO:hdx_ahcd:Downloading file:ftp://ftp.cdc.gov/pub/Health_Statistics/NCHS/namcs_public_use_files/namcs75.exe for year:1975
>>> pp.pprint(gen)
defaultdict(<class 'dict'>,
            {   1973: {   'generator': <generator object get_generator_by_year at 0x7fe4b5fa9e08>,
                          'source_file_info': {
                          'url': 'ftp://ftp.cdc.gov/pub/Health_Statistics/NCHS/namcs_public_use_files/namcs73.exe',
                          'year': '73',
                          'zip_file_name': 'namcs73.exe'}},
                1975: {   'generator': <generator object get_generator_by_year at 0x7fe4b45e7e60>,
                          'source_file_info': {
                          'url': 'ftp://ftp.cdc.gov/pub/Health_Statistics/NCHS/namcs_public_use_files/namcs75.exe',
                          'year': '75',
                          'zip_file_name': 'namcs75.exe'}}})
```
> Case 4: Forcefully download and then export the processed NAMCS data for
        multiple years (say, 1973 and 1975) in separate csv files.
```sh
>>> gen =  get_cleaned_data_by_year(year= (1973,1975), force_download=True, do_export=True)
INFO:hdx_ahcd:Downloading file:ftp://ftp.cdc.gov/pub/Health_Statistics/NCHS/namcs_public_use_files/namcs73.exe for year:1973
INFO:hdx_ahcd:Finished writing to the file /home/velotio/.hdx_ahcd/data/1973_NAMCS_CONVERTED.csv
INFO:hdx_ahcd:Downloading file:ftp://ftp.cdc.gov/pub/Health_Statistics/NCHS/namcs_public_use_files/namcs75.exe for year:1975
INFO:hdx_ahcd:Finished writing to the file /home/velotio/.hdx_ahcd/data/
1975_NAMCS_CONVERTED.csv
>>> pp.pprint(gen)
defaultdict(<class 'dict'>,
        {   1973: {   'file_name': '/home/velotio/.hdx_ahcd/data/1973_NAMCS_CONVERTED.csv',
                      'generator': <generator object get_generator_by_year at 0x7fe4b6480150>,
                      'source_file_info': {
                      'url': 'ftp://ftp.cdc.gov/pub/Health_Statistics/NCHS/namcs_public_use_files/namcs73.exe',
                      'year': '73',
                      'zip_file_name': 'namcs73.exe'}},
            1975: {   'file_name': '/home/velotio/.hdx_ahcd/data/1975_NAMCS_CONVERTED.csv',
                      'generator': <generator object get_generator_by_year at 0x7fe4b45e7e60>,
                      'source_file_info': {   '
                      url': 'ftp://ftp.cdc.gov/pub/Health_Statistics
                      /NCHS/namcs_public_use_files/namcs75.exe',
                      'year': '75',
                      'zip_file_name': 'namcs75.exe'}}})
```
> Case 5: Process the provided NAMCS data set file. In this case file name is
        assumed to follow "YEAR_NAMCS" format.
```sh
>>> gen = get_cleaned_data_by_year(file_name="/var/tmp/2015_NAMCS")
>>> pp.pprint(gen)
defaultdict(<class 'dict'>,
            {   2015: {   'generator': <generator object get_generator_by_year at 0x7f7ba17acf68>,
                          'source_file_info': {   'url': 'ftp://ftp.cdc.gov/pub/Health_Statistics/NCHS/Datasets/NAMCS/namcs2015.zip',
                                                  'year': '15',
                                                  'zip_file_name': 'namcs2015.zip'}}})
>>> pp.pprint(next(gen.get(2015).get("generator")))
{   'age': 23725,
    'month_of_visit': 'October',
    'patient_visit_weight': 414200.0481,
    'physician_diagnoses': ['723.10', '719.41', '729.50', 'V50.80', 'V00.009'],
    'sex': 'Female',
    'source_file_ID': '2015_NAMCS',
    'source_file_row': 1,
    'year_of_visit': '2015'}
```
### Uninstall
-----
To uninstall you can use either
```sh
easy_install -m hdx_ahcd
```
or
```sh
pip uninstall hdx_ahcd
```
### Scope
-----
 - Support for NHAMCS data set to be added in subsequent releases.
 - Unsupported years due to missing data sets on CDC server.
    - 1974, 1982, 1983, 1984, 1986, 1987, 1988, 1991


CHANGELOG.md
Changelog
=========


(unreleased)
------------
- Restructuring_and_refactoring_package. [Rishabh Parate]
- Merge pull request #1 from humandx/feat/download-and-extract-NAMCS.
  [Aditya Kedia]

- Added NAMCS extractor and converter for public dataset
- Added visit weight for all years. [Rishabh Parate]
- Added mappings for year 1995 and 1996. [Rishabh Parate]
- Added mapping for year 1992 to 1998. [Rishabh Parate]
- Fixed issue with relative import. [Rishabh Parate]
- Removed .pyc files. [Rishabh Parate]
- Changed package name to hdx_ahcd. [Rishabh Parate]
- Merge branch 'master' into feat/download-and-extract-NAMCS. [rishabh-
  velotio]
- Update README with AHCD description. [Scott Wong]
- Minor code fixes encountered in testing. [Rishabh Parate]
- Added traceback deatils for exeption and tests for use_regex parameter
  of decorator enforce_type. [Rishabh Parate]
- Fixed typos. [Rishabh Parate]
- Added use_regex parameter for decorator enforce_type. [Rishabh Parate]
- Added NAMCS extractor and converter for public dataset. [Rishabh
  Parate]
- Initial commit. [David Li-Bland]


hdx_ahcd/api.py
# -*- coding: utf-8 -*-
"""
Module to expose API for processing NAMCS dataset file(s).
"""
# Other modules
from hdx_ahcd.controllers.namcs_processors import NAMCSProcessor as \
    __Processor

# 3rd party modules
# -N/A


def get_cleaned_data_by_year(**kwargs):
    """
    API to process NAMCS dataset file(s).

    Parameters:
        **kwargs (:class:`dict`) : Following are permissible parameters.
            year (:class:`int` or :class:`tuple` or :class:`list`): NAMCS year.
            file_name (:class:`str`): NAMCS dataset file name.
            do_validation (:class:`bool`): If to perform validation
                on `year` and `file_name`. *Default** :const:`True`.
            do_export (:class:`bool`): Output translated  data into csv file.
            *Default** :const:`False`.
            force_download (:class:`bool`): Whether to force download
                NAMCS raw dataset file even if data set file exists locally.
                *Default** :const:`False`.
    Returns:
        :class:`defaultdict`: Dictionary containing generator of converted
            NAMCS patient case data for given year along with source file info.
            Further if `do_export` is True, it returns the absolute path of csv
            file where the data is exported.
    Usage:

    Case 1: Downloading NAMCS data for single year (say, 1973). If the file is
        already present in the downloaded_files then process the downloaded
        file.
    >>> import pprint
    >>> from hdx_ahcd.api import get_cleaned_data_by_year
    >>> pp =pprint.PrettyPrinter(indent=4)
    >>> gen =  get_cleaned_data_by_year(year=1973)
    >>> pp.pprint(gen)
    defaultdict(<class 'dict'>,
            {   1973: {   'generator': <generator object
            get_generator_by_year at 0x7fe4b6480150>,
                          'source_file_info': {
                          'url': 'ftp://ftp.cdc.gov/pub/Health_Statistics/NCHS
                          /namcs_public_use_files/namcs73.exe',
                          'year': '73',
                          'zip_file_name': 'namcs73.exe'}}})

    Case 2: Downloading NAMCS data for multiple years (say, 1973 and 1975).
        If the file is already present in the downloaded_files then process the
        downloaded file.
    >>> gen =  get_cleaned_data_by_year(year= (1973,1975))
    >>> pp.pprint(gen)
    defaultdict(<class 'dict'>,
                {   1973: {   'generator': <generator object
                get_generator_by_year at 0x7fe4b5fa9e08>,
                              'source_file_info': {
                              'url': 'ftp://ftp.cdc.gov/pub/Health_Statistics
                              /NCHS/namcs_public_use_files/namcs73.exe',
                              'year': '73',
                              'zip_file_name': 'namcs73.exe'}},
                    1975: {   'generator': <generator object
                    get_generator_by_year at 0x7fe4b45e7e60>,
                              'source_file_info': {
                              'url': 'ftp://ftp.cdc.gov/pub/Health_Statistics
                              /NCHS/namcs_public_use_files/namcs75.exe',
                              'year': '75',
                              'zip_file_name': 'namcs75.exe'}}})
    >>> pp.pprint(next(gen.get(1973).get('generator')))
    {   'age': 22889,
        'month_of_visit': 'June',
        'patient_visit_weight': 13479.0,
        'physician_diagnoses': ['470.0', 'V03.2', ''],
        'sex': 'Female',
        'source_file_ID': '1973_NAMCS',
        'source_file_row': 1,
        'year_of_visit': '1973'}
    >>> pp.pprint(next(gen.get(1975).get('generator')))
    Case 3: Forcefully download and then process the NAMCS data for multiple
        years (say, 1973 and 1975).
    >>> gen =  get_cleaned_data_by_year(year= (1973,1975), force_download=True)
    >>> pp.pprint(gen)
    defaultdict(<class 'dict'>,
                {   1973: {   'generator': <generator object
                get_generator_by_year at 0x7fe4b5fa9e08>,
                              'source_file_info': {
                              'url': 'ftp://ftp.cdc.gov/pub/Health_Statistics
                              /NCHS/namcs_public_use_files/namcs73.exe',
                              'year': '73',
                              'zip_file_name': 'namcs73.exe'}},
                    1975: {   'generator': <generator object
                    get_generator_by_year at 0x7fe4b45e7e60>,
                              'source_file_info': {
                              'url': 'ftp://ftp.cdc.gov/pub/Health_Statistics
                              /NCHS/namcs_public_use_files/namcs75.exe',
                              'year': '75',
                              'zip_file_name': 'namcs75.exe'}}})

    Case 4: Forcefully download and then export the processed NAMCS data for
        multiple years (say, 1973 and 1975) in separate csv files.
    >>> gen =  get_cleaned_data_by_year(year= (1973,1975), force_download=True,
        do_export=True)
    >>> pp.pprint(gen)
    Case 5: Forcefully download and then export the processed NAMCS data for all
            years.
    >>> gen =  get_cleaned_data_by_year(force_download=True)
    >>> pp.pprint(gen)
    Case 6: Process the provided NAMCS data set file. In this case file name is
        assumed to follow "YEAR_NAMCS" format.
    >>>gen = get_cleaned_data_by_year(file_name="/var/tmp/1973_NAMCS")
    >>> pp.pprint(gen)
    defaultdict(<class 'dict'>,
                {   1973: {   'generator': <generator object get_generator_by_year at 0x7f7ba17ac8e0>,
                              'source_file_info': {   'url': 'ftp://ftp.cdc.gov/pub/Health_Statistics/NCHS/namcs_public_use_files/namcs73.exe',
                                                      'year': '73',
                                                      'zip_file_name': 'namcs73.exe'}}})
    >>> gen = get_cleaned_data_by_year(file_name="/var/tmp/2015_NAMCS")
    >>> pp.pprint(gen)
    defaultdict(<class 'dict'>,
                {   2015: {   'generator': <generator object get_generator_by_year at 0x7f7ba17acf68>,
                              'source_file_info': {   'url': 'ftp://ftp.cdc.gov/pub/Health_Statistics/NCHS/Datasets/NAMCS/namcs2015.zip',
                                                      'year': '15',
                                                      'zip_file_name': 'namcs2015.zip'}}})
    >>> pp.pprint(next(gen.get(2015).get("generator")))
    {   'age': 23725,
        'month_of_visit': 'October',
        'patient_visit_weight': 414200.0481,
        'physician_diagnoses': ['723.10', '719.41', '729.50', 'V50.80', 'V00.009'],
        'sex': 'Female',
        'source_file_ID': '2015_NAMCS',
        'source_file_row': 1,
        'year_of_visit': '2015'}

    Case 7: Attempting to process the file for which data is not available on
        CDC FTP server.
    >>> gen = get_cleaned_data_by_year(year=1991)
    ERROR:hdx_ahcd:Year 1991 is not valid year, please specify valid years,valid years are :[1973, 1975, 1976, 1977, 1978, 1979, 1980, 1981, 1985, 1989, 1990, 1992, 1993, 1994, 1995, 1996, 1997, 1998, 1999, 2000, 2001, 2002, 2003, 2004, 2005, 2006, 2007, 2008, 2009, 2010, 2011, 2012, 2013, 2014, 2015]

    Case 8: Attempting to process a file which does not exists in local machine.
    >>> gen = get_cleaned_data_by_year(file_name="/var/tmp/2015")
    ERROR:hdx_ahcd:NAMCS dataset file:/var/tmp/2015 doesn't exist
    """
    return __Processor().execute(**kwargs)
